engines/director/lingo/xlibs/popupmenuxobj.cpp

```diff
@@ -149,26 +152,72 @@ void PopUpMenuXObj::close(int type) {
 }
 
 
-PopUpMenuXObject::PopUpMenuXObject(ObjectType ObjectType) :Object<PopUpMenuXObject>("PopMenu") {
+PopUpMenuXObject::PopUpMenuXObject(ObjectType ObjectType) : Object<PopUpMenuXObject>("PopMenu") {
 	_objType = ObjectType;
 }
 
 void PopUpMenuXObj::m_new(int nargs) {
-	g_lingo->printSTUBWithArglist("PopUpMenuXObj::m_new", nargs);
-	g_lingo->dropStack(nargs);
+	PopUpMenuXObject *me = static_cast<PopUpMenuXObject *>(g_lingo->_state->me.u.obj);
+	Datum menuId = g_lingo->pop();
+	Datum menuList = g_lingo->pop();
+
+	new Graphics::MacPopUp(menuId.asInt(), g_director->_wm->getScreenBounds(), g_director->_wm, menuList.u.s->c_str());
+	me->_menuId = menuId.asInt();
+
 	g_lingo->push(g_lingo->_state->me);
 }
 
+void PopUpMenuXObj::m_popNum(int nargs) {
+	PopUpMenuXObject *me = static_cast<PopUpMenuXObject *>(g_lingo->_state->me.u.obj);
+
+	Datum itemNum = g_lingo->pop();
+	Datum top = g_lingo->pop();
+	Datum left = g_lingo->pop();
+
+	// Convert window coordinates to screen coordinates
+	Common::Rect windowRect = g_director->getCurrentWindow()->getInnerDimensions();
+	int screenTop = top.asInt() + windowRect.top - 1;
+	int screenLeft = left.asInt() + windowRect.left - 1;
+
+	Graphics::MacPopUp *menu = static_cast<Graphics::MacPopUp *>(g_director->_wm->getMenu(me->_menuId));
+	int selected = menu->drawAndSelectMenu(screenLeft, screenTop, itemNum.asInt());
+	g_lingo->push(Datum(selected));
+}
+
+void PopUpMenuXObj::m_popText(int nargs) {
+	PopUpMenuXObject *me = static_cast<PopUpMenuXObject *>(g_lingo->_state->me.u.obj);
+
+	Datum itemNum = g_lingo->pop();
+	Datum top = g_lingo->pop();
+	Datum left = g_lingo->pop();
+
+	// Convert window coordinates to screen coordinates
+	Common::Rect windowRect = g_director->getCurrentWindow()->getInnerDimensions();
+	int screenTop = top.asInt() + windowRect.top - 1;
+	int screenLeft = left.asInt() + windowRect.left - 1;
+
+	Graphics::MacPopUp *menu = static_cast<Graphics::MacPopUp *>(g_director->_wm->getMenu(me->_menuId));
+	int selected = menu->drawAndSelectMenu(screenLeft, screenTop, itemNum.asInt());
+	Common::String text = menu->getItemText(selected);
+
+	g_lingo->push(Datum(text));
+}
+
+void PopUpMenuXObj::m_smart(int nargs){
+	PopUpMenuXObject *me = static_cast<PopUpMenuXObject *>(g_lingo->_state->me.u.obj);
+	Datum isSmart = g_lingo->pop();
+
+	Graphics::MacPopUp *menu = static_cast<Graphics::MacPopUp *>(g_director->_wm->getMenu(me->_menuId));
+	menu->setSmart(isSmart.asInt() == 1);
```

Comment:
This is an incorrect comparison. The correct one is `isSmart.asInt() != 0`, but please test and make sure with the original. But once again, it is more readable to write:

```C++
bool isSmart = g_lingo->pop().asInt() != 0;`
```